### DISCUSSION TITLE
Incorporating TSV element as layer_changing element

### DISCUSSION THREAD

**Discussion by Qvadi** [iqm-finland/KQCircuits#73] [Q&A] [answered]
Hi,
I also noticed that the TSVstandard element is not part of the layer_changing_element list. Can this also be part of the waveguide_composite element class? Would be super useful. I tried adding it and playing with the code a bit but the problem I encounter is mainly coming from the TSV element not having ports. Can it be simply added with co-incident ports, same as the base ref_point?

**Comment by iqmtestd** [comment]
Hi,

Yes, I agree,  the same idea has passed my mind too.

The usual way to change faces with `WaveguideComposite` is to use the `face_id="2b1"` Node parameter. This works fine automatically using the `FlipChipConnectorRf` element. For example, you may use this `(-200, 0),(0, 0, {"face_id": "2b1" }), (200, 0)` node_list "manually"  in  `WaveguideComposite` to see it working:

![image](https://github.com/iqm-finland/KQCircuits/assets/83572540/7fba6b8b-d809-4443-be3e-4a2e9b1f7b16)

I think the `face_id="1b1"` Node argument, could do the same thing but using a TSV element. We'll see what we can do about this.

**Comment by Qvadi** [comment]
Hi team KQC, any workarounds that you can suggest?

**Reply by qpavsmi** [reply, marked as answer]
Here is what I managed to get working using the "Standard TSV element". In `defaults.py` enable the element to be selected in the GUI "Edit node" tool by adding `'TsvStandard'` to both `node_editor_valid_elements` and `node_editor_layer_changing_elements`. Then in `elements/tsvs/tsv_standard.py` add the following lines at the end of the `build` function:

```
        self.add_port("{}_port".format(self.face_ids[0]), pya.DPoint(0, 0))
        self.add_port("{}_port".format(self.face_ids[1]), pya.DPoint(0, 0))
```

I get something like this:

![image](https://github.com/iqm-finland/KQCircuits/assets/109152229/405b64ea-750d-4fec-bf4b-6fb5dee1b634)

A literal hole in the wire, so maybe not something that we want to fabricate.

You might want to create an element of your own into a new file (use `FlipChipConnectorRf` as an example code) that defines a hole in the "through_silicon_via" layer and some geometry in the "base_metal_gap" layer supporting it. Then add the two ports I copy-pasted above in the locations that you want so that it can be as part of the WaveguideComposite.

Hope this helps.

**Reply by Qvadi** [reply]
Thanks for the suggestion. It works. I am building on the flip_chip connector module to realize something similar with TSVs.
